Makefile

```diff
@@ -5,15 +5,51 @@
 # Please see GNU make's documentation if unfamiliar: https://www.gnu.org/software/make/manual/html_node/
 .PHONY: test build e2e dist clean format lint check site
 
+# Include versions of tools we build on-demand
+include Tools.mk
+
 # This should be driven by automation and result in N.N.N, not vN.N.N
 VERSION   ?= dev
 build_dir := build
 dist_dir  := dist
 
-# Build the path relating to the current runtime (goos,goarch)
-goos   := $(shell go env GOOS)
-goarch := $(shell go env GOARCH)
-goexe  := $(shell go env GOEXE)
+# This selects the goroot to use in the following priority order:
+# 1. ${GOROOT}          - Ex actions/setup-go
+# 2. ${GOROOT_1_17_X64} - Ex GitHub Actions runner
+# 3. $(go env GOROOT)   - Implicit from the go binary in the path
+#
+# There may be multiple GOROOT variables, so pick the one matching go.mod.
+go_release           := $(shell sed -ne 's/^go //gp' go.mod)
+# https://github.com/actions/runner/blob/master/src/Runner.Common/Constants.cs
+github_runner_arch   := $(if $(findstring $(shell uname -m),x86_64),X64,ARM64)
+goroot_github_env    := $(GOROOT_$(subst .,_,$(go_release))_$(github_runner_arch))
+# This works around missing variables on macOS via naming convention.
+# Ex. /Users/runner/hostedtoolcache/go/1.17.1/x64
+# Remove this after actions/virtual-environments#4156 is solved.
+goroot_github_cache  := $(lastword $(shell ls -d $(RUNNER_TOOL_CACHE)/go/$(go_release)*/$(github_runner_arch) 2>/dev/null))
+goroot_path          := $(shell go env GOROOT 2>/dev/null)
+goroot               := $(firstword $(GOROOT) $(goroot_github_env) $(goroot_github_cache) $(goroot_path))
+
+# Ensure POSIX-style GOROOT even in Windows, to support PATH updates in bash.
+ifdef COMSPEC
+goroot := $(shell cygpath $(goroot))
+endif
+
+# We must ensure `go` executes with GOROOT and PATH variables exported:
+# * GOROOT ensures versions don't conflict with /usr/local/go or c:\Go
+# * PATH ensures tools like golint can fork and execute the correct go binary.
+#
+# We may be using a very old version of Make (ex. 3.81 on macOS). This means we
```

Comment:
this took longer because I had to learn about floor features of make, and how things like exporting variables are very hard given the floor version installed on OS/x. Rather than force all devs to upgrade Make, this bears with the glitches and hopefully documents why.